Feature/mirror gitlab repos to GitHub

RobjS opened the pull request:
This PR:

* Adds a ruby script that loops through all the repos hosted within a group on a specified GitLab instance, and mirrors as private repos in a specified GitHub org
* Adds a github action that sets up and runs that script once every 12 hours (using secrets that will be added to this repo)

To test this locally, you can use act: https://github.com/nektos/act

When testing locally, it's important to edit the `mirror-plugins` script slightly so it doesn't mirror all 700+ plugin repos! You can do that by modifying l.58-60, so it reads more like:

```
projects = Gitlab.group_projects(ENV["GITLAB_WORDPRESS_PLUGINS_GROUP_ID"], per_page: 10)

projects.each do |project|
...
```

That would just check the first 10. They key thing is to change `projects.auto_paginate` to `projects.each`, so you don't get multiple batches.

Then copy the `.env.example` to `.env`, populate the username/token values (the relevant secrets are available in 1password) and run:

```
 act -j mirror-gitlab-plugins --secret-file .env
 ```

After all the setup output, it should report that plugins are either up-to-date, exist on GitHub and need updating or (if you hit a repo that doesn't yet exist on GitHub) need creating and then updating.

evilstreak commented on .ruby-version:
```diff
@@ -0,0 +1 @@
+2.6.3
```
You'd typically want to use a more recent version for a new project, unless you have a restriction for some reason. The 2.6 branch is up to 2.6.8 now, though is only in security maintenance mode. The 2.7 and 3.0 branches are both in normal maintenance mode.

Source: https://www.ruby-lang.org/en/downloads/branches/ and https://www.ruby-lang.org/en/downloads/releases/

RobjS replied:
Thanks, updated to 2.7.2.

evilstreak commented on .env.example:
```diff
@@ -0,0 +1,22 @@
+# env.example
```
This file is really well documented 👍🏻

evilstreak commented on mirror-plugins:
```diff
@@ -0,0 +1,84 @@
+Dotenv.load
+
+Gitlab.configure do |config|
+  config.endpoint = ENV["GITLAB_API_ENDPOINT"]
```
I strongly recommend using `#fetch` to get an entry from a Hash. If the key doesn't exist `#[]` will return `nil` and `#fetch` will raise an error. In this situation, a `nil` will break unexpectedly at some point later, potentially causing more trouble, and definitely making it harder to track down the source of the issue.

RobjS replied:
Thanks, I didn't know that difference in behaviour. I've switched all the `ENV` calls to `ENV.fetch('VAR_NAME')` now.

evilstreak commented on mirror-plugins:
```diff
@@ -0,0 +1,84 @@
+  url.insert(8, credentials)
+end
+
+def mirror_gitlab_repo(project, create = false)
```
I recommend avoiding boolean arguments that change the behaviour of the method. They increase complexity and the chance of error when calling it. In this case I'd prefer to define two methods. Depending on how the implementation works, that could be:

- a method to do just the sync and a method to do just the creation (eg `sync_mirror` and `create_mirror`); or
- a method to do just the sync and a method to do create and sync (eg `mirror` and `create_and_mirror`)

RobjS replied:
I've split this out into two methods (`create_github_repo` and `update_github_mirror`) now.

evilstreak commented on README.md:
````diff
@@ -0,0 +1,34 @@
+act -j mirror-gitlab-plugins --secret-file .env
+```
+
+NOTE: this will actually start the mirroring process off for real, and there are over 700 plugins to be mirrored! So you may want to tweak lines 58 & 60 of the `mirror-plugins` script to reduce that number, e.g.:
````
It's worth considering a dry run flag for your script, so that it's easier to test it's working without actually making any changes. I'm not sure how that'd work if you're calling it via `act`!

evilstreak reviewed (approved):
It looks like it'll work! I haven't checked out the code to run it locally, so I'm just going off a reading.

I've dropped a few small comments on specific lines. My one big recommendation that would make this a lot more Rubyish (IMO) would be to use a class to wrap up all your project methods in `mirror-plugins`. In general, if you're calling `foo(thing)` and `bar(thing)` a lot, you might want a `Thing` class with `#foo` and `#bar` methods.

Doing this would also have the advantage of letting you memoise the `Gitlab.commits` call. I don't know if that's a call that triggers network requests (but I assume it is) and I don't know if it memoises the results itself (but I assume it doesn't). If those both go how I assume, you probably don't want to make the call twice for the same arguments if you can avoid it. Memoising with your current approach would either need you to combine the methods so you can use a local variable, or build a lookup table to cache results in. With a class you can memoise using an instance variable, in a very common Ruby pattern:

```ruby
def gitlab_commits
  @gitlab_commits ||= Gitlab.commits(…)
end
```

If you're interested in writing this as a class and want some help let me know!

RobjS commented:
Thanks @evilstreak, that was really useful - especially the pointer about how lots of functions passing the same argument are a pointer towards the need for a class.

I think I've addressed all the points except the `--dry-run` flag, so this should be ready for another look now.
